File: nonLeetCodeQ/addString.java
/*Implement add: 

"1000", "-100" == "900"

*/
	String addString(String s1, String s2)
	{
		s1=s1.trim();
		s2=s2.trim();
		int m = s1.length(), n=s2.length();
		if(m==0) return s2;
		if(n==0) return s1;
		boolean f1=(s1.charAt(0)=='-'), f2=(s2.charAt(0)=='-');

		if(s1.charAt(0)=='+'||s1.charAt(0)=='-') s1=s1.substring(1);
		if(s2.charAt(0)=='+'||s2.charAt(0)=='-') s2=s2.substring(1);

		String re = "";
		if((f1&&f2)||(!f1 && !f2))
		{
			re = helper(s1,s2, true);
			if(f1) re = "-"+re;
		}
		else
		{
			re = (f1)?helper(s2,s1,false):helper(s1,s2,false);
		}
		return re;
	}

	String helper(String s1, String s2, boolean add)
	{
		String re = "";
		int i=s1.length()-1, j=s2.length()-1;
		int carry=0;
		while(i>=0 || j>=0)
		{
			if(i<0)
			{
				int b = (!add)?('0'-s2.charAt(j)):(s2.charAt(j)-'0');
				int r = b+carry;
				if(r>=0)
				{
					re = Integer.toString(r%10)+re;
					carry = r/10;
				}
				else
				{
					re = Integer.toString(r+10)+re;
					carry = -1;
				}
				j--;
			}
			else if(j<0)
			{
				int a = s1.charAt(i--)-'0';
				int r = a+carry;
				if(r>=0)
				{
					re = Integer.toString(r%10)+re;
					carry = r/10;
				}
				else
				{
					re = Integer.toString(r+10)+re;
					carry = -1;
				}
			}
			else
			{
				int a = s1.charAt(i)-'0';
				int b = (!add)?('0'-s2.charAt(j)):(s2.charAt(j)-'0');
				int r = a+b+carry;
				if(r>=0)
				{
					re = Integer.toString(r%10)+re;
					carry = r/10;
				}
				else
				{
					re = Integer.toString(r+10)+re;
					carry = -1;
				}
				i--;
				j--;
			}
		}
		while(re.length()!=0 && re.charAt(0)=='0') re=re.substring(1);
		if(re.length()==0) return "0";
		if(carry==-1) re="-"+re;
		return re;
	}

	public static void main(String[] args) {
		System.out.println(new Solution().addString("-1000","-1000"));
	}
}

File: nonLeetCodeQ/addString.py
import sys, copy, Queue, math, time
from ojClass import *

class Solution:

    def addString(self, str1, str2):
        print str1,str2,
        m,n=len(str1.strip()),len(str2.strip())
        if m==0:
            return str2
        if n==0:
            return str1
        f1,f2=False, False
        if str1[0] in ['+','-']:
            f1=str1[0]=='-'
            str1=str1[1::]
        if str2[0] in ['+','-']:
            f2=str2[0]=='-'
            str2=str2[1::]
        re=""
        m,n=len(str1.strip()),len(str2.strip())
        if (f1 and f2) or (not f1 and not f2):
            re=helper(str1,str2,True)
            if f1 and re!='0':
                re='-'+re
        else:
            if check(str1,str2):
                re=helper(str1,str2, False)
                if f1 and re!='0':
                    re='-'+re
            else:
                re=helper(str2,str1,False)
                if f2 and re!='0':
                    re='-'+re

        return re
def check(str1, str2):
    m,n=len(str1),len(str2)
    if m>n:
        return True
    if m<n:
        return False
    i,j=0,0
    while i<m:
        if int(str1[i])>int(str2[j]):
            return True
        if int(str1[i])<int(str2[j]):
            return False
        i+=1
        j+=1
    return True

def helper(str1, str2, add):
    re=""
    carry=0

    i,j=len(str1)-1,len(str2)-1
    while i>=0 or j>=0:
        val=carry
        if i>=0:
            val+=int(str1[i])
            i-=1
        if j>=0:
            if not add:
                val-=int(str2[j])
                j-=1
            else:
                val+=int(str2[j])
                j-=1
        if val>=0:
            re=str(val%10)+re
            carry=val/10
        else:
            carry=0
            while val<0:
                carry-=1
                val+=10
            re=str(val%10)+re
    i=0
    while i<len(re) and re[i]=='0':
        i+=1
    if i==len(re):
        re="0"
    else:
        re=re[i::]

    if carry>0:
        re=str(carry)+re
    if carry<0:
        re='-'+str(-carry)+re
    return re


if __name__=="__main__":

    print Solution().addString("100","-1000")
    print Solution().addString("","-1000")
    print Solution().addString("","1000")
    print Solution().addString("100","1000")
    print Solution().addString("-100","-1000")
    print Solution().addString("-1000","100")
    print Solution().addString("-1000","-100")
    print Solution().addString("-1000","1000")
